RFC-0005-structured-kernel-definitions.md

```diff
@@ -0,0 +1,323 @@
+
+* **Major format change to native_functions.yaml**: We could do this. We
+  are choosing not to in order to reduce the degrees of freedom in what
+  changes we make to `native_functions.yaml`.
+
+## Summary
+
+* Core
+
+    * Define a new API for shape checking functions. Shape checking
+      functions are called from generated code that implements
+      functional/inplace/out variants of functions.
+
+    * Define a new API for kernels. Static kernel API does not do shape
+      checking, output is always preallocated at correct size. This API
+      is public and a suitable entrypoint for static runtime. Static
+      kernels are called from generated code like above.
+
+    * Code generate meta functions from shape checking functions to
+      provide public API for running shape computations without any
+      kernel.
+
+    * Generated code is augmented to do version counter bumps and view
+      tracking
+
+* Extensions
+
+    * Add a new dispatch key (name tbd) which contains shape checking,
+      device guard, version counter bump. External backend
+      implementations of PyTorch operators which do not explicitly opt
+      out of this dispatch key will get this logic applied to them. This
+      key is skipped for core operators, as this checking logic is fused
+      directly into the backend kernel. External backends can also opt
+      to fuse this logic in.
+
+## Details
+
+Let’s suppose you want to write a new operator from scratch. The first
+thing you will do is add a native_functions.yaml declaration for it.
+Let’s take `upsample_nearest1d` as the example for this post.  You will
+write an entry like this, describing the *functional* version of this
+operator:
+
+```
+- func: upsample_nearest1d(Tensor self, int[1] output_size, float? scales=None) -> Tensor
+  dispatch:
+    CPU: upsample_nearest1d_cpu
+    CUDA: upsample_nearest1d_cuda
+```
+
+Because you are a conscientious implementor, you also want to provide an
+out= variant of this function. This version gets a separate entry:
+
+```
+- func: upsample_nearest1d.out(Tensor self, int[1] output_size, float? scales=None, *, Tensor(a!) out) -> Tensor(a!)
+  dispatch:
+    CPU: upsample_nearest1d_out_cpu
+    CUDA: upsample_nearest1d_out_cuda
+```
+
+Ordinarily, these two declarations would cause function signatures for
+`upsample_nearest1d_cpu`, etc., to be generated into NativeFunctions.h,
+and then you would go ahead and implement them.  However, we propose a
+new *structured* format for writing kernels. We’ll do this by marking
+the out version of this operator as structured and deleting dispatch
+entries from the functional version (the functional operator is
+*implicitly* associated with the out-of-place version in the same way
+derivatives.yaml schemas are matched with their functional/inplace/out
+variants):
+
+```
+- func: upsample_nearest1d(Tensor self, int[1] output_size, float? scales=None) -> Tensor
+  # [NEW] dispatches for this function are omitted
```

Comment:
hm maybe we should make this explicit by also adding `structured: True` here, to make the information more local and easier to read. Otherwise you have to look at other entries in native_functions.yaml to know what this entry is actually doing. Omitting dispatch keys is already a valid kernel definition even if it's not structured, so by just looking at this one it would be ambigious.